modules/ml/src/main/java/org/apache/ignite/ml/selection/cv/AbstractCrossValidation.java

```diff
@@ -0,0 +1,476 @@
+import org.apache.ignite.ml.selection.paramgrid.ParamGrid;
+import org.apache.ignite.ml.selection.paramgrid.ParameterSetGenerator;
+import org.apache.ignite.ml.selection.paramgrid.RandomStrategy;
+import org.apache.ignite.ml.selection.scoring.cursor.LabelPairCursor;
+import org.apache.ignite.ml.selection.scoring.metric.Metric;
+import org.apache.ignite.ml.selection.split.mapper.SHA256UniformMapper;
+import org.apache.ignite.ml.selection.split.mapper.UniformMapper;
+import org.apache.ignite.ml.trainers.DatasetTrainer;
+import org.apache.ignite.ml.util.genetic.Chromosome;
+import org.apache.ignite.ml.util.genetic.GeneticAlgorithm;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Cross validation score calculator. Cross validation is an approach that allows to avoid overfitting that is made the
+ * following way: the training set is split into k smaller sets. The following procedure is followed for each of the k
+ * “folds”:
+ * <ul>
+ * <li>A model is trained using k-1 of the folds as training data;</li>
+ * <li>the resulting model is validated on the remaining part of the data (i.e., it is used as a test set to compute
+ * a performance measure such as accuracy).</li>
+ * </ul>
+ *
+ * @param <M> Type of model.
+ * @param <L> Type of a label (truth or prediction).
+ * @param <K> Type of a key in {@code upstream} data.
+ * @param <V> Type of a value in {@code upstream} data.
+ */
+public abstract class AbstractCrossValidation<M extends IgniteModel<Vector, L>, L, K, V> {
+    /** Learning environment builder. */
+    protected LearningEnvironmentBuilder envBuilder = LearningEnvironmentBuilder.defaultBuilder();
+
+    /** Learning Environment. */
+    protected LearningEnvironment environment = envBuilder.buildForTrainer();
+
+    /** Trainer. */
+    protected DatasetTrainer<M, L> trainer;
+
+    /** Pipeline. */
+    protected Pipeline<K, V, Integer, Double> pipeline;
+
+    /** Metric. */
+    protected Metric<L> metric;
+
+    /** Preprocessor. */
+    protected Preprocessor<K, V> preprocessor;
+
+    /** Filter. */
+    protected IgniteBiPredicate<K, V> filter = (k, v) -> true;
+
+    /** Amount of folds. */
+    protected int amountOfFolds;
+
+    /** Parts. */
+    protected int parts;
+
+    /** Parameter grid. */
+    protected ParamGrid paramGrid;
+
+    /** Execution over the pipeline or the chain of preprocessors and separate trainer, otherwise. */
+    protected boolean isRunningOnPipeline = true;
+
+    /** Mapper. */
+    protected UniformMapper<K, V> mapper = new SHA256UniformMapper<>();
+
+    /**
+     * Finds the best set of hyperparameters based on parameter search strategy.
+     */
+    public CrossValidationResult tuneHyperParamterers() {
+        HyperParameterTuningStrategy hyperParamTuningStgy = paramGrid.getHyperParameterTuningStrategy();
+
+        if (hyperParamTuningStgy instanceof BruteForceStrategy) return scoreBruteForceHyperparameterOptimiztion();
```

Comment:
missprint - optimiztion/optimization

Reply:
Why you didn't implement scoreBruteForceHyperparameterOptimiztion() method in BruteForceStrategy?